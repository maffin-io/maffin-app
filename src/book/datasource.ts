import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';
import { useSWRConfig } from 'swr';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

// TODO: Need to mutate the accounts SWR key so Root is shown!
async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
