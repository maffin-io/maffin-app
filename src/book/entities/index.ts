export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
