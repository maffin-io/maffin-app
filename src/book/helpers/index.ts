export { currencyToSymbol } from './currencyToSymbol';
export { isInvestment, isLiability, isAsset } from './accountType';
