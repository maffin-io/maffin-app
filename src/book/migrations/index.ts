import { AddCommodityName1697688851711 } from './1697688851711-AddCommodityName';
import { AddBankConfig1717405195056 } from './1717405195056-AddBankConfig';

export const MIGRATIONS = [AddBankConfig1717405195056, AddCommodityName1697688851711];
