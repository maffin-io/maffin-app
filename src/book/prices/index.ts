export { default as PriceDBMap } from './PriceDBMap';
