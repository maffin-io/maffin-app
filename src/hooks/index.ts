export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
