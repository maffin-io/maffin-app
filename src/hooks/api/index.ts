import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
