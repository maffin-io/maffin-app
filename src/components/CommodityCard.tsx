import React from 'react';
import Link from 'next/link';
import {
  BiLineChart,
  BiMoney,
  BiSolidGhost,
} from 'react-icons/bi';

import { Tooltip, UpgradeTooltip } from '@/components/tooltips';
import { useCommodity, useMainCurrency, usePrices } from '@/hooks/api';
import Loading from '@/components/Loading';
import { IS_PAID_PLAN } from '@/helpers/env';

export type CommodityCardProps = {
  guid: string;
};

export default function CommodityCard({
  guid,
}: CommodityCardProps): JSX.Element {
  const { data: commodity } = useCommodity(guid);
  const { data: mainCurrency } = useMainCurrency();
  const { data: prices } = usePrices({ from: commodity });

  if (!commodity || !prices) {
    return <Loading />;
  }

  let iconComponent = <BiLineChart className="mr-1" />;
  let tooltipContent = (
    <div>
      <p>
        No prices found for this commodity! This means that accounts
        using this commodity don&apos;t have an exchange rate to convert
        to your main currency and we will convert 1 to 1 for now.
      </p>

      <p className="mt-1">
        To fix this, make sure the symbol can be found in

        {' '}
        <Link
          href={`https://finance.yahoo.com/quote/${commodity.mnemonic}`}
          target="blank"
        >
          Yahoo
        </Link>
        . If not, consider adding this commodity
        as &quot;OTHER&quot; type instead and add prices manually.
      </p>
    </div>
  );

  if (commodity.namespace === 'CURRENCY') {
    iconComponent = <BiMoney className="mr-1" />;
    tooltipContent = (
      <div>
        <p>
          No prices found for the
          {' '}
          {`${commodity.mnemonic}.${mainCurrency?.mnemonic}`}
          {' '}
          pair! This means that accounts using this currency don&apos;
          have an exchange rate to convert to your main currency and we
          will convert 1 to 1 for now.
        </p>

        <p className="mt-1">
          To fix this, make sure the currency can be found in
          {' '}
          <Link
            href={`https://finance.yahoo.com/quote/${commodity.mnemonic}${mainCurrency?.mnemonic}=X`}
            target="blank"
          >
            Yahoo
          </Link>
          . If it does not exist, it means you will have to track prices for this currency
          manually.
        </p>
      </div>
    );
  } else if (commodity.namespace === 'OTHER') {
    tooltipContent = (
      <div>
        <p>
          No prices found for this commodity! This means that accounts
          using this commodity don&apos;t have an exchange rate to convert
          to your main currency and we will convert 1 to 1 for now.
        </p>

        <p className="mt-1">
          To fix this, make sure to add prices manually for this commodity
        </p>
      </div>
    );
    iconComponent = <BiSolidGhost className="mr-1" />;
  }

  let priceAvailable = false;
  if (commodity.namespace === 'CURRENCY') {
    const price = prices.getPrice(commodity.mnemonic, mainCurrency?.mnemonic || '');
    priceAvailable = price?.guid !== 'missing_price';
  } else {
    const price = prices.getInvestmentPrice(commodity.mnemonic);
    priceAvailable = price?.guid !== 'missing_price';
  }

  return (
    <>
      <Link
        key={commodity.guid}
        href={`/dashboard/commodities/${commodity.guid}`}
        className="card col-span-2 cursor-pointer hover:shadow-lg text-slate-400 dark:hover:text-white"
      >
        <div className="flex items-center">
          {iconComponent}
          {commodity.mnemonic}
          {
            !priceAvailable
            && (
              <div
                className="badge warning ml-auto"
                data-tooltip-id={`missing-price-${guid}`}
              >
                !
              </div>
            )
          }
        </div>
      </Link>

      {
        !priceAvailable
        && (
          IS_PAID_PLAN
            ? (
              <Tooltip
                id={`missing-price-${guid}`}
                className="!w-1/6 !text-xs"
                clickable
              >
                {tooltipContent}
              </Tooltip>
            )
            : (
              <UpgradeTooltip
                id={`missing-price-${guid}`}
                message="Non paid users need to add prices manually. Consider upgrading for getting a bunch of commodities synced automatically for you!"
              />
            )
        )
      }
    </>
  );
}
