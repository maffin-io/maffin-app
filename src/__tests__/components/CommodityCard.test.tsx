import React from 'react';
import { render, screen } from '@testing-library/react';
import type { UseQueryResult } from '@tanstack/react-query';

import * as apiHooks from '@/hooks/api';
import { PriceDBMap } from '@/book/prices';
import type { Commodity, Price } from '@/book/entities';
import CommodityCard from '@/components/CommodityCard';

jest.mock('@/hooks/api');

jest.mock('@/components/Loading', () => jest.fn(
  () => <div data-testid="Loading" />,
));

jest.mock('@/helpers/env', () => ({
  __esModule: true,
  get IS_PAID_PLAN() {
    return true;
  },
}));

describe('CommodityCard', () => {
  let eur: Commodity;
  beforeEach(() => {
    jest.spyOn(apiHooks, 'useCommodity').mockReturnValue({ data: undefined } as UseQueryResult<Commodity>);
    jest.spyOn(apiHooks, 'usePrices').mockReturnValue({ data: undefined } as UseQueryResult<PriceDBMap>);

    eur = {
      mnemonic: 'EUR',
      namespace: 'CURRENCY',
    } as Commodity;
    jest.spyOn(apiHooks, 'useMainCurrency').mockReturnValue({
      data: eur,
    } as UseQueryResult<Commodity>);
  });

  it('returns loading when no commodity', () => {
    jest.spyOn(apiHooks, 'usePrices').mockReturnValue({ data: {} } as UseQueryResult<PriceDBMap>);

    render(<CommodityCard guid="guid" />);

    screen.getByTestId('Loading');
  });

  it('returns loading when no prices', () => {
    jest.spyOn(apiHooks, 'useCommodity').mockReturnValue({ data: {} } as UseQueryResult<Commodity>);

    render(<CommodityCard guid="guid" />);

    screen.getByTestId('Loading');
  });

  it('renders as expected without warning when prices for CURRENCY', () => {
    jest.spyOn(apiHooks, 'useCommodity').mockReturnValue({
      data: {
        mnemonic: 'EUR',
        namespace: 'CURRENCY',
      },
    } as UseQueryResult<Commodity>);

    jest.spyOn(apiHooks, 'usePrices').mockReturnValue({
      data: {
        getPrice: (() => ({ guid: 'price' } as Price)) as PriceDBMap['getPrice'],
        getInvestmentPrice: (() => ({ guid: 'price' } as Price)) as PriceDBMap['getInvestmentPrice'],
      } as PriceDBMap,
    } as UseQueryResult<PriceDBMap>);

    const { container } = render(<CommodityCard guid="guid" />);

    screen.getByText('EUR');
    expect(container).toMatchSnapshot();
  });

  it.each([
    'STOCK', 'MUTUAL', 'OTHER',
  ])('renders as expected without warning when prices for non CURRENCY', (namespace) => {
    jest.spyOn(apiHooks, 'useCommodity').mockReturnValue({
      data: {
        mnemonic: 'MNE',
        namespace,
      },
    } as UseQueryResult<Commodity>);

    jest.spyOn(apiHooks, 'usePrices').mockReturnValue({
      data: {
        getPrice: (() => ({ guid: 'price' } as Price)) as PriceDBMap['getPrice'],
        getInvestmentPrice: (() => ({ guid: 'price' } as Price)) as PriceDBMap['getInvestmentPrice'],
      } as PriceDBMap,
    } as UseQueryResult<PriceDBMap>);

    render(<CommodityCard guid="guid" />);

    screen.getByText('MNE');
    expect(screen.queryByText('!')).toBeNull();
  });

  it.each([
    'CURRENCY', 'STOCK', 'MUTUAL', 'OTHER',
  ])('renders as expected without warning when prices for %s', (namespace) => {
    jest.spyOn(apiHooks, 'useCommodity').mockReturnValue({
      data: {
        mnemonic: 'MNE',
        namespace,
      },
    } as UseQueryResult<Commodity>);

    jest.spyOn(apiHooks, 'usePrices').mockReturnValue({
      data: {
        getPrice: (() => ({ guid: 'missing_price' } as Price)) as PriceDBMap['getPrice'],
        getInvestmentPrice: (() => ({ guid: 'missing_price' } as Price)) as PriceDBMap['getInvestmentPrice'],
      } as PriceDBMap,
    } as UseQueryResult<PriceDBMap>);

    render(<CommodityCard guid="guid" />);

    screen.getByText('MNE');
    screen.getByText('!');
  });
});
