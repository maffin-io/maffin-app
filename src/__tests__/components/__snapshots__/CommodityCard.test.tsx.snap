// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommodityCard renders as expected without warning when prices for CURRENCY 1`] = `
<div>
  <a
    class="card col-span-2 cursor-pointer hover:shadow-lg text-slate-400 dark:hover:text-white"
    href="/dashboard/commodities/undefined"
  >
    <div
      class="flex items-center"
    >
      <svg
        class="mr-1"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 4H3a1 1 0 0 0-1 1v14a1 1 0 0 0 1 1h18a1 1 0 0 0 1-1V5a1 1 0 0 0-1-1zm-1 11a3 3 0 0 0-3 3H7a3 3 0 0 0-3-3V9a3 3 0 0 0 3-3h10a3 3 0 0 0 3 3v6z"
        />
        <path
          d="M12 8c-2.206 0-4 1.794-4 4s1.794 4 4 4 4-1.794 4-4-1.794-4-4-4zm0 6c-1.103 0-2-.897-2-2s.897-2 2-2 2 .897 2 2-.897 2-2 2z"
        />
      </svg>
      EUR
    </div>
  </a>
</div>
`;
